Kernel/Devices/VMWareSVGADevice.cpp

```diff
@@ -0,0 +1,280 @@
+
+    // TODO: The reference drivers "fakes" 3D support here. Should we do the same?
+    set_register(IORegister::ConfigDone, true);
+
+    // TODO: Test Interrupts
+    dbgln("VMWareSVGA: Set mode {}x{}x{}, pitch {}", width, height, bpp, m_framebuffer_pitch);
+
+    m_mode_set = true;
+}
+
+int VMWareSVGADevice::ioctl(FileDescription&, unsigned request, FlatPtr arg)
+{
+    REQUIRE_PROMISE(video);
+    switch (request) {
+    case FB_IOCTL_GET_SIZE_IN_BYTES: {
+        auto* out = (size_t*)arg;
+        size_t value = m_framebuffer_size;
+        if (!copy_to_user(out, &value)) {
+            return -EFAULT;
+        }
+        return 0;
+    }
+    case FB_IOCTL_GET_BUFFER: {
+        auto* index = (int*)arg;
+        int value = 0;
+        if (!copy_to_user(index, &value))
+            return -EFAULT;
+        return 0;
+    }
+    case FB_IOCTL_SET_BUFFER: {
+        // We do not support double buffering
+        return -EINVAL;
+    }
+    case FB_IOCTL_GET_RESOLUTION: {
+        auto* user_resolution = (FBResolution*)arg;
+        FBResolution resolution;
+        resolution.pitch = m_framebuffer_pitch;
+        resolution.width = m_framebuffer_width;
+        resolution.height = m_framebuffer_height;
+        if (!copy_to_user(user_resolution, &resolution))
+            return -EFAULT;
+        return 0;
+    }
+    case FB_IOCTL_SET_RESOLUTION: {
+        auto* user_resolution = (FBResolution*)arg;
+        FBResolution resolution;
+        if (!copy_from_user(&resolution, user_resolution))
+            return -EFAULT;
+        set_mode(resolution.width, resolution.height, 32);
+        resolution.width = m_framebuffer_width;
+        resolution.height = m_framebuffer_height;
+        resolution.pitch = m_framebuffer_pitch;
+        if (!copy_to_user(user_resolution, &resolution))
+            return -EFAULT;
+        return 0;
+    }
+
+    default: {
+        return -EINVAL;
+    }
+    }
+}
+
+KResultOr<Region*> VMWareSVGADevice::mmap(Process& process, FileDescription&, const Range& range, u64 offset, int prot, bool shared)
+{
+    REQUIRE_PROMISE(video);
+    if (!shared)
+        return ENODEV;
+    if (offset != 0)
+        return ENXIO;
+    if (range.size() != page_round_up(m_framebuffer_size))
+        return EOVERFLOW;
+
+    auto vmobject = AnonymousVMObject::create_for_physical_range(m_framebuffer_address_physical, m_framebuffer_size);
+    if (!vmobject)
+        return ENOMEM;
+    return process.space().allocate_region_with_vmobject(
+        range,
+        vmobject.release_nonnull(),
+        0,
+        "VMWare SVGA Framebuffer",
+        prot,
+        shared);
+}
+
+String VMWareSVGADevice::device_name() const
+{
+    return String::formatted("fb{}", minor());
+}
+
+void VMWareSVGADevice::handle_irq(const RegisterState&)
+{
+    // Read and clear the interrupts
+    u16 irq_port = m_io_base + static_cast<u32>(Ports::IRQStatus);
+    u32 irqs = IO::in32(irq_port);
+    IO::out32(irq_port, irqs);
+
+    VERIFY(irqs);
```

Comment:
If this shares an interrupt with another device, this can cause a crash as it'll read in 0 if the interrupt is for the other device and not this one. I'd recommend returning with a debug message saying something along the lines of `Interrupt not for this device or spurious`